Intermittent test failure: test_analysis_results_view_data_1_Complete

This has been happening a lot locally and in CI:

[https://app.circleci.com/pipelines/github/mozilla/experimenter/7336/workflows/908e1f22-1b8a-470b-99d5-5b478921b7ed/jobs/18274](https://app.circleci.com/pipelines/github/mozilla/experimenter/7336/workflows/908e1f22-1b8a-470b-99d5-5b478921b7ed/jobs/18274)

```

=================================== FAILURES ===================================
_______ TestVisualizationView.test_analysis_results_view_data_1_Complete _______
[gw0] linux -- Python 3.9.0 /usr/local/bin/python

a = (<experimenter.visualization.tests.api.test_views.TestVisualizationView testMethod=test_analysis_results_view_data_1_Complete>,)

@wraps(func)
 def standalone_func(*a):
> return func(*(a + p.args), **p.kwargs)

/usr/local/lib/python3.9/site-packages/parameterized/parameterized.py:530: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.9/unittest/mock.py:1337: in patched
 return func(*newargs, **newkeywargs)
/usr/local/lib/python3.9/site-packages/parameterized/parameterized.py:74: in dummy_func
 return orgfunc(*args, **kwargs)
/usr/local/lib/python3.9/unittest/mock.py:1337: in patched
 return func(*newargs, **newkeywargs)
experimenter/visualization/tests/api/test_views.py:224: in test_analysis_results_view_data
 response = self.client.get(
/usr/local/lib/python3.9/site-packages/django/test/client.py:733: in get
 response = super().get(path, data=data, secure=secure, **extra)
/usr/local/lib/python3.9/site-packages/django/test/client.py:393: in get
 return self.generic('GET', path, secure=secure, **{
/usr/local/lib/python3.9/site-packages/django/test/client.py:470: in generic
 return self.request(**r)
/usr/local/lib/python3.9/site-packages/django/test/client.py:710: in request
 self.check_exception(response)
/usr/local/lib/python3.9/site-packages/django/test/client.py:571: in check_exception
 raise exc_value
/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py:47: in inner
 response = get_response(request)
/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py:179: in _get_response
 response = wrapped_callback(request, *callback_args, **callback_kwargs)
/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py:54: in wrapped_view
 return view_func(*args, **kwargs)
/usr/local/lib/python3.9/site-packages/django/views/generic/base.py:70: in view
 return self.dispatch(request, **args,** *kwargs)
/usr/local/lib/python3.9/site-packages/rest_framework/views.py:509: in dispatch
 response = self.handle_exception(exc)
/usr/local/lib/python3.9/site-packages/rest_framework/views.py:469: in handle_exception
 self.raise_uncaught_exception(exc)
/usr/local/lib/python3.9/site-packages/rest_framework/views.py:480: in raise_uncaught_exception
 raise exc
/usr/local/lib/python3.9/site-packages/rest_framework/views.py:506: in dispatch
 response = handler(request, **args,** *kwargs)
/usr/local/lib/python3.9/site-packages/rest_framework/decorators.py:50: in handler
 return func(*args, **kwargs)
experimenter/visualization/api/v3/views.py:182: in analysis_results_view
 data = process_data_for_consumption(data, experiment)
experimenter/visualization/api/v3/views.py:116: in process_data_for_consumption
 append_conversion_count(results, primary_metrics_set)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

results = {'control': {'branch_data': {'identity': {'absolute': {'lower': 10, 'point': 12, 'upper': 13}, 'difference': {}, 'perc...e': {'lower': 10, 'point': 12, 'upper': 13}, 'relative_uplift': {}, 'significance': 'positive'}}, 'is_control': False}}
primary_metrics_set = {'configurable-cohesive-circuit_ever_used'}

def append_conversion_count(results, primary_metrics_set):
 for branch in results:
 branch_data = results[branch][BRANCH_DATA]
 for primary_metric in primary_metrics_set:
 population_count = branch_data[Metric.USER_COUNT][BranchComparison.ABSOLUTE][
 "point"
 ]
> conversion_percent = branch_data[primary_metric][BranchComparison.ABSOLUTE][
 "point"
 ]
E KeyError: 'configurable-cohesive-circuit_ever_used'

experimenter/visualization/api/v3/views.py:104: KeyError

```

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-619)
┆Issue Number: EXP-619

--- a/app/experimenter/visualization/tests/api/test_views.py
+++ b/app/experimenter/visualization/tests/api/test_views.py
@@ -6,7 +6,7 @@ from django.test import TestCase, override_settings
 from django.urls import reverse
 from parameterized import parameterized
 
-from experimenter.experiments.constants import ExperimentConstants
+from experimenter.experiments.models import NimbusExperiment
 from experimenter.experiments.tests.factories import (
     NimbusExperimentFactory,
     NimbusProbeSetFactory,
@@ -18,8 +18,8 @@ from experimenter.visualization.api.v3.views import Significance
 class TestVisualizationView(TestCase):
     @parameterized.expand(
         [
-            ExperimentConstants.STATUS_ACCEPTED,
-            ExperimentConstants.STATUS_COMPLETE,
+            NimbusExperiment.Status.ACCEPTED,
+            NimbusExperiment.Status.COMPLETE,
         ]
     )
     @patch("django.core.files.storage.default_storage.exists")
@@ -49,20 +49,17 @@ class TestVisualizationView(TestCase):
             json_data,
         )
 
-    def add_probe_set_data(self, data, formatted_data, probe_sets):
-        if len(probe_sets) == 0:
-            return
-
+    def add_probe_set_data(self, data, formatted_data, primary_probe_set):
         range_data = {
             "point": 4,
             "upper": 8,
             "lower": 2,
         }
         branches = ["control", "variant"]
-        probe_set = f"{probe_sets[0].slug}_ever_used"
+        primary_metric = f"{primary_probe_set.slug}_ever_used"
 
         for branch in branches:
-            formatted_data[branch]["branch_data"][probe_set] = {
+            formatted_data[branch]["branch_data"][primary_metric] = {
                 "absolute": {**range_data, **{"count": 48}},
                 "difference": {},
                 "relative_uplift": {},
@@ -71,7 +68,7 @@ class TestVisualizationView(TestCase):
                 {
                     **range_data,
                     **{
-                        "metric": probe_set,
+                        "metric": primary_metric,
                         "branch": branch,
                         "statistic": "binomial",
                     },
@@ -80,8 +77,8 @@ class TestVisualizationView(TestCase):
 
     @parameterized.expand(
         [
-            ExperimentConstants.STATUS_ACCEPTED,
-            ExperimentConstants.STATUS_COMPLETE,
+            NimbusExperiment.Status.ACCEPTED,
+            NimbusExperiment.Status.COMPLETE,
         ]
     )
     @patch("django.core.files.storage.default_storage.open")
@@ -206,9 +203,9 @@ class TestVisualizationView(TestCase):
 
         mock_open.return_value = File()
         mock_exists.return_value = True
-        probe_set_primary = NimbusProbeSetFactory.create()
+        primary_probe_set = NimbusProbeSetFactory.create()
         experiment = NimbusExperimentFactory.create_with_status(
-            target_status=status, probe_sets=[probe_set_primary]
+            target_status=status, probe_sets=[primary_probe_set]
         )
         experiment.probe_sets.add(
             NimbusProbeSetFactory.create(),
@@ -218,7 +215,7 @@ class TestVisualizationView(TestCase):
         self.add_probe_set_data(
             DATA_WITHOUT_POPULATION_PERCENTAGE,
             FORMATTED_DATA_WITH_POPULATION_PERCENTAGE,
-            experiment.probe_sets.all(),
+            primary_probe_set,
         )
 
         response = self.client.get(
@@ -230,7 +227,7 @@ class TestVisualizationView(TestCase):
         json_data = json.loads(response.content)
         self.assertEqual(FULL_DATA, json_data)
 
-    @parameterized.expand([ExperimentConstants.STATUS_ACCEPTED])
+    @parameterized.expand([NimbusExperiment.Status.ACCEPTED])
     def test_analysis_results_view_no_experiment(self, status):
         user_email = "user@example.com"
         response = self.client.get(

